### ISSUE TITLE
Add public query object model API draft

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#12425]
A variety of SQL transformation use-cases are currently impossible, or hard to implement both internally and for users, because jOOQ's query object model:

- Is package private and has no public API (that's only a user problem)
- Is mutable and does not have a very unform transformation API

A new public query object model API could help, here:

- [x] A new API could be offered, e.g. in `org.jooq.impl.QOM.*` (this would allow package private access to internals).
- [x] That API could profit from new JDK features including:
  - [x] Interfaces could be sealed (JDK 17: https://openjdk.java.net/jeps/409). Some observations:
    - [x] ~Does pattern matching work if final classes are package private?~ Yes, it does. If there's a 1:1 relationship  between public interface and private implementation, then exhaustiveness can still be checked based on the interface alone. However, if a `permits` clause includes public and private types, then sealed types are useless for public API consumers. Thus, there must be one public type for each implementation in `QOM`
    - [x] The `permits` clause requires its types to be visible. Hence, the API cannot be moved to `org.jooq` if its goal is to permit `org.jooq.impl` implementation types! Perhaps we should move the implementation to `org.jooq`, instead? This would clean up a few other things...
    - [x] Some bottom-up sealed type refactoring should be possible already. E.g. auxiliary types like `MCompareCondition` can be sealed, which already adds value.
    - [x] ecj's Java 17 beta support seems to have a bug that produces invalid byte code. I'll have to postpone sealed type support for now
    - [x] ~In fact, if `MQueryPart` types are sealed, their `QueryPart` counterparts need to be, as well, and the entire type hierarchy becomes very complex. This will require a few iterations to get right...~ It was decided not to have separate type hierarchies.
    - [x] Work will be continued in https://github.com/jOOQ/jOOQ/issues/12772
  - [x] ~Implementations could be records (JDK 16: https://openjdk.java.net/jeps/395). This is tricky, because records can't extend abstract base classes. To be verified~. Impractical given the complex type hierarchies
  - [x] The API can be generated from our new API generation tooling
- [x] The API includes
  - [x] `QueryPart::$replace` to transform an expression tree and form a new one from replacements (without modifying the existing one!)
    - [x] We need an extension that prevents entering subtrees, just like with `traverse()`
  - [x] `QueryPart::$traverse` iterate the tree and do generic stuff with each element (all the others can use this)
    - [x] Just like with `Stream.collect()`, there should be `$traverse` convenience methods that help build a `Traverser` (possibly to be renamed)
    - [x] Add a `Traverser` (better name needed) API similar to the `Collector` API, which can be composed
    - [x] Add a `Traversers` utility to construct common traversers:
      - [x] `Traversers.collecting(Collector)` to turn a JDK `Collector` into a `Traverser`
      - [x] `Traversers.containing(QueryPart)` allowing to search for the existence of elements in an expression tree
      - [x] `Traversers.findingAny(Predicate<? super QueryPart>)` and `Traversers.findingAll(Predicate<? super QueryPart>)` producing an `Optional` or `List` of elements that satisfy a predicate (`Traversers.findingAll(p)` is the same as `Traversers.collecting(Collectors.filtering(p, Collectors.toList()))`
      - [x] The current implementation is stateful, i.e. a `Traverser` can be used only once (unlike `Collectors`). It's probably better to remove this state from the `Traverser`?
  - [x] All sorts of marker interfaces, e.g. for all subquery types, etc. Some examples:
    - [x] `QOM.CompareCondition`
    - [x] `QOM.CombinedCondition`
    - [x] ~`QOM.SubqueryScope` (?) Could be useful to declare query parts that produce a new scope level (e.g. `Select`, others?)~
  - [x] Utility interfaces
    - [x] `UEmpty` are `QOM` types that don't have any contents. Some of these may be composed of other `QOM` types, but they don't traverse their children, e.g. a `TableFieldImpl` doesn't traverse its `Table` or `Name`
    - [x] `UTransient` are `QOM` types that are never part of an expression tree, but are used only to generate SQL
    - [x] `UNotYetImplemented` is a marker for things that aren't supported (yet)
    - [x] `UProxy` is an `QueryPart` that represents another `QueryPart`. Traversal should never be communicated to user code

### Design

The `org.jooq.impl.QOM` API can be implemented in two ways:

- As a separate type hierarchy where each `Q <: QueryPart` has a `M <: QOM.MQueryPart` equivalent, with these sub options
  - `Q <: M`
  - `M <: Q`
  - `Q` and `M` have no subtype relationships, and grant access to each other via `Q.model()` and `M.dsl()` methods
- In the same `QueryPart` type hierarchy with common types being reused (e.g. `org.jooq.Field`), and subtyped in `org.jooq.impl.QOM` where needed (e.g. to open up internals, such as `Eq`, `Ne`, etc.)

Ultimately, the separate type hierarchy idea was rejected. While it helps cleanly separate DSL building and model transformation concerns, the concerns aren't necessarily so different. For example, when introspecting an `Eq` expression tree for transformation purposes, one would want to be able to build new expression trees via the DSL API again. Separate type hierarchies are a lot of work, lead to unsafe casts, multiple inheritance edge cases and much more. The price of "polluting" the `QueryPart` type hierarchy with methods like `traverse()` is relatively low.

### Prerequisites

There are many prerequisites, which will be implemented first.

**Required prerequisites**:

- [x] Refactor associativity logic (avoiding unnecessary parentheses) between `Expression` and `CombinedCondition`
- [x] Review all function parameter names (given that we now have effectively named parameters via the generated model API accessors)

**Optional prerequisites** (making things much easier):

- [x] https://github.com/jOOQ/jOOQ/issues/10710
- [x] https://github.com/jOOQ/jOOQ/issues/12427
- [x] Move more functionality to the API generator
- [x] Split existing implementations into several types
  - [x] New `GroupField` types for `GROUPING SETS` and `CUBE` (we already have `ROLLUP`)
  - [x] All the window functions
  - [x] All the aggregate functions (including `MIN`, `MAX`, etc)

**Rejected**

- [x] ~Define for each `NegatableCondition` if negation should be a condition feature or an external one (e.g. `NOT EXISTS` or `NOT (EXISTS)`)~ Each negatable condition now ships as two `QOM` types, e.g. `In`, `NotIn`. This is to avoid the potential confusion where `A NOT <op> B` is not the same thing as `NOT A <op> B`, which is the case for the `NULL` predicate, for example
- [x] ~The `VisitListener` feature can finally be deprecated~. The `VisitListener` is still useful, and now even more powerful, as users can replace subtrees while rendering
- [x] https://github.com/jOOQ/jOOQ/issues/11241

### Follow up work

- [x] Manual
  - [x] A section about the API
  - [x] A section containing a few transformation and other usage examples
- [x] Blog and documentation about common transformations (moved to an internal issue)

### Clean up

- [x] Review all "operators" with `arg1`, `arg2`, `arg3`, to see if some of them deserve a better name. Solution: The numbered operands stay there, such that we can use them if we need access by index in utilities, but additional methods with more meaningful names (such as `dividend` and `divisor`) can be generated and bridged to `arg1`, `arg2`
- [x] `UProxy` type that can't be replaced / traversed directly, but acts as if it were its proxied object.
  - [x] `traverse()` should (probably?) ignore the proxy and directly reach for the proxied object
  - [x] `replace()` should (probably?) behave ordinarily, i.e. if the proxied object is replaced, it needs to be wrapped in a new proxy again.
- [x] Make sure `replace()` is optimal, and that we're not doing too much work
- [x] Deprecations
  - [x] ~`org.jooq.True` and `org.jooq.False` won't be required anymore if we have `QOM` types for these~. We can still keep these, and we'll need a bridge type too, to give access to the internal types from the sealed classes...
- [x] Remove the `Condition` suffix from conditions (?)
- [x] Refactor existing methods to use the new transformation APIs instead
  - [x] `transform` (various occasions)
    - [x] `And` and `Or` (used to transform `ROWNUM`)
    - [x] `Transform` utility
  - [x] `Tools::traverseXYZ`
    - [x] `traverseConditions`
    - [x] `traverseSelectList`
    - [x] `traverseOrderBy`
  - [x] `Eq`, `Ne`, `Gt`, `Ge`, `Lt`, `Le` usages
  - [x] `And`, `Or` usages
- [x] `AbstractCondition.unwrapNot` could be removed (?)
- [x] Resolve the inconsistency between existing `XML` types and `Xml` ones
`noCondition()` could be the default)
- [x] Decide whether `QOM` is the right name, or just `Model`
- [x] Decide whether to go with the `$xyz()` property accessor naming strategy or JavaBeans style `getXyz()` (potential conflicts with existing methods)
- [x] `UOperatorN` types (with `$arg1()`, `$arg2()` style arguments) should offer alternatives with actual names (?)
- [x] Tests
  - [x] API
    - [x] A type that is tagged `UTransient` must never be found in any expression tree via `QueryPart.findAny` (this could be a test that is part of the parser test suite)
    - [x] `U` types should be package private, `QOM` types and enums public
    - [x] Add a test that checks if all accessor methods start with the same prefix (currently `$`, might change)
    - [x] Check if QOM types don't have any `Boolean` accessors (prefer enums, instead, where `null` semantics is more clear), e.g. `CreateSequence::$cache`
    - [x] Check if all QOM type properties can be accessed via a `$xyz()` method
    - [x] Make sure all QOM API is annotated with `@Nullable`or `@NotNull`
      - [x] Correct `@NotNull` behaviour can be checked by calling all the accessor methods
      - [x] Assert that some types are never `@Nullable` (e.g. `MList`)
  - [x] Behaviour
    - [x] Implement systematic `traverse()` and `replace()` tests for each QOM type
- [x] Copy
  - [x] Generate setters that produce copies changing only one value
  - [x] ~Think about "mutating setters", i.e. an alternative API that allows for using a builder-style API to construct the copy without copying the object on each setter call. This is especially useful e.g. for `SELECT` (probably not too useful otherwise, so could be implemented ad-hoc, not generated everywhere)~. Out of scope for now.
- [x] For backwards compatibility reasons, we need to keep all XML and JSON type names in upper case, not Xml or Json
- [x] ~The `Tuple2` type makes me wonder if we should finally let jOOQ depend on jOOλ, see also https://github.com/jOOQ/jOOQ/issues/4420, https://github.com/jOOQ/jOOQ/issues/3184~
- [x] The generated `equals()` methods might behave subtly differently from the previous ones: We're now comparing components rather than the generated string. We should detect as many regressions as possible! One workaround is to avoid generating `equals()` on `Attachable & QueryPart` subtypes. Some observed differences:
  - [x] It no longer depends on an attached `Configuration` (e.g. quoting of identifiers)
  - [x] `val(1)` is no longer equal to `val(1L)` in a query context (they never were equal when compared directly)
- [x] Add documentation everywhere

### Rejected

- [x] ~Remove the `Impl` suffix from statements~ (there's no compelling need to do this for now, as long as the types are internal)
- [x] ~Generate `AbstractCondition.isNullable()`~ (might be implemented later, independently of this task)

**Comment by lukaseder** [comment]
The status quo is almost ready to be shipped in jOOQ 3.16 as an experimental feature. A lot more will be done in the next version of jOOQ, see https://github.com/jOOQ/jOOQ/issues/12772
